MIPS/mip-zkapps.md

```diff
@@ -0,0 +1,961 @@
+defines a new tree in the forest, and a non-zero number defines a child of the
+most recently defined tree.
+
+Here is an example of a forest constructed from a list of account updates with
+call-depths `0 0 1 2 3 1 0` :
+
+```
+ *  *     *
+    | \
+    *  *
+    |
+    *
+    |
+    *
+```
+
+#### Updates in Account Updates
+
+An account update can contain a request to update certain fields in an account.
+Those fields are:
+
+- `app_state`
+- `delegate`
+- `verification_key`
+- `permissions`
+- `zkapp_uri`
+- `token_symbol`
+- `timing`
+- `voting_for`
+
+Note that some of these items are not in the in the `zkapp` part of an account
+(for example, the `delegate` field). A given account update can request to
+update some or all of these items.
+
+The `permissions` field of accounts is expanded to include permissions to update
+each of these items. Besides requiring permission, the `timing` field of an
+account can be set only if the account is currently untimed. The complete set of
+permissions is described.
+
+#### Preconditions in account updates
+
+Three kinds of preconditions can be checked before applying an account update:
+
+- account preconditions
+- network preconditions
+- a `validWhile` precondition
+
+The account preconditions are:
+
+- `balance` : a range of account balances
+- `nonce` : a specific nonce, or a range of nonces
+- `receipt_chain_hash` : a specific receipt chain hash
+- `delegate` : a specific public key
+- `state` : a specific application state
+- `sequence_state` : a specific sequence state
+- `proved_state` : a specific proved state
+- `is_new` : whether the account is new
+
+An account update can include some or all of these
+preconditions. These preconditions are checked against the account
+from the previous protocol state. These conditions must be true at the
+moment the transaction is included in a block, otherwise the
+transaction fails.
+
+The network preconditions are:
+
+- `snarked_ledger_hash` : a specific ledger hash
+- `blockchain_length` : a range of block heights
+- `min_window_density` : a range of window densities
+- `total_currency` : a range of amounts
+- `global_slot_since_genesis` : a range of slots
+- `staking_epoch_data` : epoch data
+- `next_epoch_data` : epoch_data
+
+where `epoch_data` contains:
+
+- a specific ledger hash
+- a range for total currency
+- a specific epoch seed
+- start and lock checkpoints
+- a range of epoch lengths
+
+Among these network precondition items, any item can be checked or ignored as a
+precondition. That also holds for the individual items in `epoch_data`. As for
+account preconditions, these items are checked against the previous protocol
+state.
+
+The initial implementation proposes that network preconditions are to be true at
+the moment the transaction is included in a block, otherwise the transaction
+fails. This failure is known to be a bit frustrating as many of the network
+preconditions . In a future hardfork, there will likely be changes to allow
```

Comment:
```suggestion
preconditions. In a future hardfork, there will likely be changes to allow
```

Reply:
verify the sentence completeness... the failure frustration is ?

maybe r/w as 

>  frustrating as many of the network preconditions are.